### PR TITLE
B 644

### PR DESCRIPTION
**Pull request by Ayelet41**
SLOs and SLIs documentation( tutorial and readme)

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -298,15 +298,15 @@ To see all messages that are being generated in Cloud Logging from your logpoint
 
 #### Monitoring Overview
 
-Cloud Monitoring ([documentation](https://cloud.google.com/monitoring/docs/)) is the go-to place to grasp real-time trends of the system based on SLI/SLO. SRE team and application development team (and even business organization team) can collaborate to set up charts on the monitoring dashboard using metrics sent from the resources and the applications. 
+Cloud Monitoring ([documentation](https://cloud.google.com/monitoring/docs/)) is the go-to place to grasp real-time trends of the system based on SLI/SLO. SRE team and application development team (and even business organization team) can collaborate to set up charts on the monitoring dashboard using metrics sent from the resources and the applications.
```
Comment (nicoleczhu) [inline docs/README.md:301]: Nit: extraneous space

File: docs/README.md [hunk 2]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
 
 ![image](./images/user-guide/35-logdet.png)
 
+### SLIs, SLOs and Burn rate Alerts
+
+Cloud operations sandbox comes with several predefined SLOs(Service level objectives), that allow us to measure our users happiness. To learn more about SLIs and SLOs [visit here.](https://cloud.google.com/blog/products/devops-sre/sre-fundamentals-slis-slas-and-slos)
```
Comment (nicoleczhu) [inline docs/README.md:446]: Nit: capitalize Operations sandbox, space after SLOs

Nit: For links, mention the page name instead of here. i.e. visit [SRE fundamentals](...)

File: docs/README.md [hunk 3]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+Cloud operations suite provides **service oriented monitoring**, that means that we are configuring SLIs, SLOs and Burning Rates Alerts for a 'service'.  
+
+The first step in order to create SLO is to **ingest the data**, for GKE services  it comes out of the box, but you can also ingest additional data and [create custom metrics.](#Monitoring-and-logs-based-metrics)
```
Comment (nicoleczhu) [inline docs/README.md:450]: Nit: Extra space after "services"

File: docs/README.md [hunk 4]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+Then we need to **define our service**, Cloud Operations Sandbox' services are already detected since Istio automatically detects and creates services for us. But to demonstrate that you can create your own services, it also deploys custom services using [Terraform](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/tree/master/terraform/monitoring).
```
Comment (nicoleczhu) [inline docs/README.md:452]: "Cloud Operations Sandbox's"

What are you intending to convey with this sentence? Istio is just a service mesh that can orchestrate telemetry, not sure it detects and creates services.

File: docs/README.md [hunk 5]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+You can find all the services under [monitoring → services → Services Overview](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/ui/svc-overview) , and you can create your own [custom service.](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/ui/define-svc)
```
Comment (nicoleczhu) [inline docs/README.md:454]: Nit: extra space after `svc-overview)`

File: docs/README.md [hunk 6]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+> **Remember**  Our scope to examine and measure our users' happiness is User journey, so in order to create the SLO you need to identify the most important ones to the business. Then we want to *identify the metrics* that are closest to the customer experience and ingest that data.  
+
+You can [configure your own SLIs and SLOs](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/ui/create-slo) for an existing service or for your own custom service.  
+
+#### Let's demonstrate that using the checkout auto-defined Istio service:
+1. In the service screen we will choose Create SLO:
+![image](./images/user-guide/39-checkout-service.png)
+2.Then we will set our SLI, we need to choose SLI type and the method(request vs window based):
+![image](./images/user-guide/42-checkoutservice-sli.png)
+3. Then we wil define our metric and we can also preview it's performance based on historical data:
```
Comment (nicoleczhu) [inline docs/README.md:507]: `wil` mispelling

File: docs/README.md [hunk 7]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+![image](./images/user-guide/41-checkoutservice-define-sli.png)
+4. Then we will configure our SLO, our target in a specific time window. We can also choose between [rolling window or a calendar window](https://sre.google/workbook/implementing-slos/):
+![image](./images/user-guide/43-set-slo.png)
+
+#### Configure Burn Rate Alerts
+
+After you created the SLO, you can create[Burn Rate Alerts](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/alerting-on-budget-burn-rate)for those.
```
Comment (nicoleczhu) [inline docs/README.md:514]: Nit: change to `After you create the SLO, create [Burn Rate Alerts]... for them`

File: docs/README.md [hunk 8]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+Several *predefined polices* are deployed as part of [Terraform](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/blob/master/terraform/monitoring/05_alerting_policies.tf). You can view them in the service screen, edit them, or [create your own](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/alerting-on-budget-burn-rate).
```
Comment (nicoleczhu) [inline docs/README.md:516]: typo: `policies`

File: docs/README.md [hunk 9]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+Let's continue with the Istio checkoutservice SLO we created [in the previous section:](#Let's-demonstrate-that-using-the-checkout-auto-defined-Istio-service:)
+
+1. In the service screen we will be able to see our new SLO and we will choose 'Create Alerting Policy'
+
+![image](./images/user-guide/46-crete-slo-burn-alert.png)
+2. Then we will want to set the alert's condition, who and how they will be notified and additional instructions:  
+![image](./images/user-guide/44-set-slo-burn-alert.png)
+3. After it will be created you could see it and incidents that might be triggered due to it in teh service screen and in the Alerting screen:
```
Comment (nicoleczhu) [inline docs/README.md:525]: typo: `the`

File: docs/README.md (hunk 3 above)
Comment (Daniel-Sanche) [inline docs/README.md:450]: nit: the , should be a period after "ingest the data". Also, you should specify what you mean by "it comes out of the box". (do you mean logs are captured?)

File: docs/README.md (hunk 4 above)
Comment (Daniel-Sanche) [inline docs/README.md:452]: you should use "you" and "your" instead of "we" and "our"

File: docs/README.md [hunk 10]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+Then we need to **define our service**, Cloud Operations Sandbox' services are already detected since Istio automatically detects and creates services for us. But to demonstrate that you can create your own services, it also deploys custom services using [Terraform](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/tree/master/terraform/monitoring).
+
+You can find all the services under [monitoring → services → Services Overview](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/ui/svc-overview) , and you can create your own [custom service.](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/ui/define-svc)
+
+![image](./images/user-guide/37-services-overview.png)  
+
+#### Services SLOs  
+
+The *predefined SLOs* are also deployed as part of [Terraform code](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/tree/master/terraform/monitoring/04_slos.tf) and currently are for the mentioned custom services, the Istio service and Rating service.
```
Comment (Daniel-Sanche) [inline docs/README.md:460]: Instead of saying the mentioned custom services, it might be a good idea to give a list (and a maybe link for each one?):
```
Cloud Ops Sandbox's Terraform deployment automatically sets up SLOs for the following services:
- ratingservice
- adservice
- ...
```

File: docs/README.md [hunk 11]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+
+![image](./images/user-guide/37-services-overview.png)  
+
+#### Services SLOs  
+
+The *predefined SLOs* are also deployed as part of [Terraform code](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/tree/master/terraform/monitoring/04_slos.tf) and currently are for the mentioned custom services, the Istio service and Rating service.  
+
+**Custom services SLOs**
+``` 
+Custom service availability SLO: 90% of all non-4XX requests within the past 30 day windowed period return with 200 OK status
```
Comment (Daniel-Sanche) [inline docs/README.md:464]: nit: Maybe reword?

> 90% of all valid (non-4xx) requests within the past 30 days return `200 OK`

Comment (Daniel-Sanche) [reply docs/README.md:464]: This comment applies to the others below as well

File: docs/README.md [hunk 12]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+```
+Rating service latency SLO: 99% of requests that return 200 OK responses return in under 175 ms
+```
+```
+Rating service's data freshness SLO: during a day 99.9% of minutes have at least 1 successful recollect API call
+```
+
+#### Configure your own SLIs and SLOs
+
+> **Remember**  Our scope to examine and measure our users' happiness is User journey, so in order to create the SLO you need to identify the most important ones to the business. Then we want to *identify the metrics* that are closest to the customer experience and ingest that data.
```
Comment (Daniel-Sanche) [inline docs/README.md:498]: This could use some re-wording. Maybe:

> **Remember** The purpose of defining SLIs and SLOs is to improve your user's experience. Take some time to identify the most important [user journeys](https://en.wikipedia.org/wiki/User_journey) around your product, and use them to define which metrics to track.

File: docs/README.md [hunk 13]
```diff
@@ -441,6 +441,90 @@ You can expand any of the messages that matches the filter to see the full stack
...
+Rating service's data freshness SLO: during a day 99.9% of minutes have at least 1 successful recollect API call
+```
+
+#### Configure your own SLIs and SLOs
+
+> **Remember**  Our scope to examine and measure our users' happiness is User journey, so in order to create the SLO you need to identify the most important ones to the business. Then we want to *identify the metrics* that are closest to the customer experience and ingest that data.  
+
+You can [configure your own SLIs and SLOs](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/ui/create-slo) for an existing service or for your own custom service.  
+
+#### Let's demonstrate that using the checkout auto-defined Istio service:
```
Comment (Daniel-Sanche) [inline docs/README.md:502]: nit: I'd prefer a different title here. Maybe "Configuring the checkout service"?

File: docs/README.md (hunk 4 above)
Comment (Ayelet41) [reply docs/README.md:452]: Rephrased

File: docs/README.md (hunk 3 above)
Comment (Ayelet41) [reply docs/README.md:450]: replaced with telemetry and dashboards. is it clearer?

File: docs/README.md (hunk 4 above)
Comment (Ayelet41) [reply docs/README.md:452]: replaced throughout the text

File: docs/README.md (hunk 10 above)
Comment (Ayelet41) [reply docs/README.md:460]: The reason I didn't add it this way, as from the code I notice that it's define for custom service,  which are defined separately. So if someone will go to the TF code and will look for it, they will not be able to find it https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/tree/master/terraform/monitoring/04_slos.tf .
I can add custom services that are defined in services.tf WDYT?

i.e. The *predefined SLOs* are also deployed as part of [Terraform code](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/tree/master/terraform/monitoring/04_slos.tf) and currently are for the mentioned custom services( defined in [services.tf](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/blob/master/terraform/monitoring/03_services.tf), the Istio service and Rating service.

File: docs/README.md (hunk 12 above)
Comment (Ayelet41) [reply docs/README.md:498]: This change the meaning a bit, and this is one of the pitfalls with our customers. 

> **Remember** The purpose of defining SLIs and SLOs is to improve your user's experience, your SLOs scope is a [User journey](https://cloud.google.com/solutions/defining-SLOs#why_slos). Therefore your first step should be to *identify the most critical User Journey(CUJ)* to your business, then *identify the metrics* that measure your customer experience as closely as possible and *ingest* that data.  

WDYT?

File: docs/README.md (hunk 11 above)
Comment (Ayelet41) [reply docs/README.md:464]: I wanted it to be aligned with the TF.
I rephrased to align with our suggested SLO format, both in TF notes and in this page accordingly.

File: docs/README.md (hunk 3 above)
Comment (daniel-sanche) [reply docs/README.md]: yeah, that looks great

File: docs/README.md (hunk 12 above)
Comment (daniel-sanche) [reply docs/README.md]: Yeah, I like that better, thanks!

**Review by daniel-sanche — APPROVED** [review]
Thanks for putting this together, LGTM!

File: docs/README.md [hunk 14]
```diff
@@ -14,7 +14,7 @@ The Hipster Shop application consists of a number of microservices, written in a
 
 ![image](./images/user-guide/1-architecture.png)
 
-**Note:** We are not endorsing this architecture as the best way to build a real online store. This application is optimized for demonstration and learning purposes.  It illustrates a large number of cloud-native technologies, uses a variety of programming languages, and provides an environment that can be explored productively with Ops Management tools.
+**Note:** You are not endorsing this architecture as the best way to build a real online store. This application is optimized for demonstration and learning purposes.  It illustrates a large number of cloud-native technologies, uses a variety of programming languages, and provides an environment that can be explored productively with Ops Management tools.
```
Comment (simonz130) [inline docs/README.md:17]: **We** (i.e. the authors/maintainers of Sandbox) are not endorsing this architecture
